src/libs/protoboard/blackboard_manager.cpp

```diff
@@ -0,0 +1,195 @@
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Library General Public License for more details.
+ *
+ *  Read the full text in the LICENSE.GPL file in the doc directory.
+ */
+
+#include "blackboard_manager.h"
+
+#include "protobuf_to_bb.h"
+
+namespace protoboard {
+
+using namespace fawkes;
+
+AbstractProtobufSender::AbstractProtobufSender(BlackboardManager *bb_mgr) : bb_manager(bb_mgr)
+{
+}
+
+AbstractProtobufSender::~AbstractProtobufSender()
+{
+}
+
+BlackboardManager::BlackboardManager(ProtobufThead *msg_handler)
+: fawkes::Thread("ProtoboardBlackboardManager", Thread::OPMODE_WAITFORWAKEUP),
+  message_handler_(msg_handler),
+  bb_receiving_interfaces_(make_receiving_interfaces_map()),
+  on_message_waker_(nullptr),
+  next_peer_idx_(0)
+{
+	set_coalesce_wakeups(true);
+}
+
+void
+BlackboardManager::set_protobuf_sender(AbstractProtobufSender *sender)
+{
+	pb_sender_.reset(sender);
+}
+
+void
+BlackboardManager::init()
+{
+	pb_sender_->init();
+	peer_iface_ = blackboard->open_for_writing<ProtobufPeerInterface>("/protoboard/peers");
+
+	on_message_waker_ = new fawkes::BlackBoardOnMessageWaker(blackboard, peer_iface_, this);
+
+	for (pb_conversion_map::value_type &c : bb_receiving_interfaces_)
+		c.second->init(blackboard, logger);
+}
+
+void
+BlackboardManager::finalize()
+{
+	delete on_message_waker_;
+	bb_receiving_interfaces_.clear();
+	pb_sender_->finalize();
+	blackboard->close(peer_iface_);
+}
+
+void
+BlackboardManager::loop()
+{
+	// Handle CreatePeer* messages
+	bool did_something = on_interface<ProtobufPeerInterface>{peer_iface_, this}
+	                       .handle_msg_types<ProtobufPeerInterface::CreatePeerMessage,
+	                                         ProtobufPeerInterface::CreatePeerLocalMessage,
+	                                         ProtobufPeerInterface::CreatePeerCryptoMessage,
+	                                         ProtobufPeerInterface::CreatePeerLocalCryptoMessage>();
+
+	// Handle sending blackboard interfaces
+	did_something |= pb_sender_->process_sending_interfaces();
+
+	// Handle receiving blackboard interfaces
+	while (message_handler_->pb_queue_incoming()) {
+		ProtobufThead::incoming_message inc = message_handler_->pb_queue_pop();
+		pb_conversion_map::iterator     it;
+
+		if ((it = bb_receiving_interfaces_.find(inc.msg->GetTypeName()))
+		    == bb_receiving_interfaces_.end()) {
+			logger->log_error(name(),
+			                  "Received message of unregistered type `%s'",
+			                  inc.msg->GetTypeName().c_str());
+			continue;
+		}
+		try {
+			it->second->handle(inc.msg);
+		} catch (std::exception &e) {
+			logger->log_error(name(),
+			                  "Exception while handling %s: %s",
+			                  inc.msg->GetTypeName().c_str(),
+			                  e.what());
+		}
+
+		did_something = true;
+	}
+
+	if (!did_something)
+		// Thread woke up, but nothing was handled
+		logger->log_warn(name(), "Spurious wakeup. WTF?");
```

Comment:
Why do we need this? It still occurs occasionally and the warning is more irritating than helpful. Can't we just remove this? Then, we could also get rid of the `did_something` flag

Reply:
True. It only happens because we're using the `BlackBoardOnMessageWaker`, which just triggers a `loop()` instead of implementing an interface-specific event handler. That would be part of the internal redesign I was talking about. For the time being we can just get rid of the warning and hope that eventually someone comes around to do clean event handling.